// react-advance-jalaali-datepicker/test-app/src/App.tsx
import React, { ClassAttributes, FC, InputHTMLAttributes, useState } from "react";
import {
  DatePicker,
  DateTimePicker,
  DateRangePicker,
  DateTimeRangePicker
} from "react-advance-jalaali-datepicker";

const App = () => {
  const [state, setState] = useState({
    valueDate: "",
    valueDateTime: "",
    valueRDateStart: "",
    valueRDateEnd: "",
    valueRDateTimeStart: "",
    valueRDateTimeEnd: ""
  })
  
  const change = (unix: any, formatted: any) => {
    console.log('unix', unix); // returns timestamp of the selected value, for example.
    console.log('formatted', formatted); // returns the selected value in the format you've entered, forexample, "تاریخ: 1396/02/24 ساعت: 18:30".
    setState(state => ({
      ...state,
      valueDate: formatted
    }))
  }

  const DatePickerInput: FC<React.InputHTMLAttributes<{}>> = (props) => {
    return <input className="popo" {...props} />;
  }

  return (
      <div className="App">
        <button
          onClick={() => {
            setState({
              valueDate: "1398/06/25",
              valueDateTime: "تاریخ: 1396/08/26 ساعت: 18:30",
              valueRDateStart: "1398/06/25",
              valueRDateEnd: "1398/06/25",
              valueRDateTimeStart: "تاریخ: 1396/08/26 ساعت: 18:30",
              valueRDateTimeEnd: "تاریخ: 1396/08/26 ساعت: 18:30"
            });
          }}
        >
          reset
        </button>
        <div dir="rtl" className="datePicker">
          <DatePicker
            inputComponent={DatePickerInput}
            placeholder="انتخاب تاریخ"
            format="jYYYY/jMM/jDD"
            onChange={change}
            cancelOnBackgroundClick={true}
            id="datePicker"
            preSelected={state.valueDate}
            controlValue={true}
          />
          <DateTimePicker
            placeholder="انتخاب تاریخ و ساعت"
            format="تاریخ: jYYYY/jMM/jDD ساعت: HH:mm"
            id="dateTimePicker"
            onChange={(unix: any, formatted: any) =>
              setState((state) => ({
                ...state,
                valueDateTime: formatted
              }))
            }
            cancelOnBackgroundClick={true}
            preSelected={state.valueDateTime}
            controlValue={true}
          />
          <DateRangePicker
            placeholderStart="تاریخ شروع"
            placeholderEnd="تاریخ پایان"
            format="jYYYY/jMM/jDD"
            onChangeStart={(unix: any, formatted: any) =>
              setState((state) => ({
                ...state,
                valueDateTime: formatted
              }))
            }
            cancelOnBackgroundClick={true}
            onChangeEnd={(unix: any, formatted: any) =>
              setState((state) => ({
                ...state,
                valueDateTime: formatted
              }))
            }
            preSelectedStart={state.valueRDateStart}
            idStart="rangePickerStart"
            idEnd="rangePickerEnd"
            controlValue={true}
          />
          <DateTimeRangePicker
            placeholderStart="تاریخ و ساعت شروع"
            placeholderEnd="تاریخ و ساعت پایان"
            format="تاریخ: jYYYY/jMM/jDD ساعت: HH:mm"
            onChangeStart={(unix: any, formatted: any) =>
              setState((state) => ({
                ...state,
                valueDateTime: formatted
              }))
            }
            preSelectedStart={state.valueRDateTimeStart}
            onChangeEnd={(unix: any, formatted: any) =>
              setState((state) => ({
                ...state,
                valueDateTime: formatted
              }))
            }
            cancelOnBackgroundClick={true}
            idStart="rangePickerStart"
            idEnd="rangePickerEnd"
            controlValue={true}
          />
        </div>
      </div>
    );
}

export default App;
