// @rjsf/material-ui/dist/material-ui.cjs.development.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@rjsf/core');
var React = require('react');
var AddIcon = require('@material-ui/icons/Add');
var IconButton = require('@material-ui/core/IconButton');
var Box = require('@material-ui/core/Box');
var Grid = require('@material-ui/core/Grid');
var Paper = require('@material-ui/core/Paper');
var utils = require('@rjsf/utils');
var TextField = require('@material-ui/core/TextField');
var Typography = require('@material-ui/core/Typography');
var ErrorIcon = require('@material-ui/icons/Error');
var List = require('@material-ui/core/List');
var ListItem = require('@material-ui/core/ListItem');
var ListItemIcon = require('@material-ui/core/ListItemIcon');
var ListItemText = require('@material-ui/core/ListItemText');
var ArrowUpwardIcon = require('@material-ui/icons/ArrowUpward');
var ArrowDownwardIcon = require('@material-ui/icons/ArrowDownward');
var RemoveIcon = require('@material-ui/icons/Remove');
var FormHelperText = require('@material-ui/core/FormHelperText');
var FormControl = require('@material-ui/core/FormControl');
var Button = require('@material-ui/core/Button');
var Divider = require('@material-ui/core/Divider');
var InputLabel = require('@material-ui/core/InputLabel');
var Input = require('@material-ui/core/Input');
var Checkbox = require('@material-ui/core/Checkbox');
var FormControlLabel = require('@material-ui/core/FormControlLabel');
var FormGroup = require('@material-ui/core/FormGroup');
var FormLabel = require('@material-ui/core/FormLabel');
var Radio = require('@material-ui/core/Radio');
var RadioGroup = require('@material-ui/core/RadioGroup');
var Slider = require('@material-ui/core/Slider');
var MenuItem = require('@material-ui/core/MenuItem');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var AddIcon__default = /*#__PURE__*/_interopDefaultLegacy(AddIcon);
var IconButton__default = /*#__PURE__*/_interopDefaultLegacy(IconButton);
var Box__default = /*#__PURE__*/_interopDefaultLegacy(Box);
var Grid__default = /*#__PURE__*/_interopDefaultLegacy(Grid);
var Paper__default = /*#__PURE__*/_interopDefaultLegacy(Paper);
var TextField__default = /*#__PURE__*/_interopDefaultLegacy(TextField);
var Typography__default = /*#__PURE__*/_interopDefaultLegacy(Typography);
var ErrorIcon__default = /*#__PURE__*/_interopDefaultLegacy(ErrorIcon);
var List__default = /*#__PURE__*/_interopDefaultLegacy(List);
var ListItem__default = /*#__PURE__*/_interopDefaultLegacy(ListItem);
var ListItemIcon__default = /*#__PURE__*/_interopDefaultLegacy(ListItemIcon);
var ListItemText__default = /*#__PURE__*/_interopDefaultLegacy(ListItemText);
var ArrowUpwardIcon__default = /*#__PURE__*/_interopDefaultLegacy(ArrowUpwardIcon);
var ArrowDownwardIcon__default = /*#__PURE__*/_interopDefaultLegacy(ArrowDownwardIcon);
var RemoveIcon__default = /*#__PURE__*/_interopDefaultLegacy(RemoveIcon);
var FormHelperText__default = /*#__PURE__*/_interopDefaultLegacy(FormHelperText);
var FormControl__default = /*#__PURE__*/_interopDefaultLegacy(FormControl);
var Button__default = /*#__PURE__*/_interopDefaultLegacy(Button);
var Divider__default = /*#__PURE__*/_interopDefaultLegacy(Divider);
var InputLabel__default = /*#__PURE__*/_interopDefaultLegacy(InputLabel);
var Input__default = /*#__PURE__*/_interopDefaultLegacy(Input);
var Checkbox__default = /*#__PURE__*/_interopDefaultLegacy(Checkbox);
var FormControlLabel__default = /*#__PURE__*/_interopDefaultLegacy(FormControlLabel);
var FormGroup__default = /*#__PURE__*/_interopDefaultLegacy(FormGroup);
var FormLabel__default = /*#__PURE__*/_interopDefaultLegacy(FormLabel);
var Radio__default = /*#__PURE__*/_interopDefaultLegacy(Radio);
var RadioGroup__default = /*#__PURE__*/_interopDefaultLegacy(RadioGroup);
var Slider__default = /*#__PURE__*/_interopDefaultLegacy(Slider);
var MenuItem__default = /*#__PURE__*/_interopDefaultLegacy(MenuItem);

const AddButton = _ref => {
  let {
    uiSchema,
    ...props
  } = _ref;
  return /*#__PURE__*/React__default["default"].createElement(IconButton__default["default"], {
    title: "Add Item",
    ...props,
    color: "primary"
  }, /*#__PURE__*/React__default["default"].createElement(AddIcon__default["default"], null));
};

const ArrayFieldItemTemplate = props => {
  const {
    children,
    disabled,
    hasToolbar,
    hasMoveDown,
    hasMoveUp,
    hasRemove,
    index,
    onDropIndexClick,
    onReorderClick,
    readonly,
    uiSchema,
    registry
  } = props;
  const {
    MoveDownButton,
    MoveUpButton,
    RemoveButton
  } = registry.templates.ButtonTemplates;
  const btnStyle = {
    flex: 1,
    paddingLeft: 6,
    paddingRight: 6,
    fontWeight: "bold",
    minWidth: 0
  };
  return /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    container: true,
    alignItems: "center"
  }, /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    item: true,
    xs: true,
    style: {
      overflow: "auto"
    }
  }, /*#__PURE__*/React__default["default"].createElement(Box__default["default"], {
    mb: 2
  }, /*#__PURE__*/React__default["default"].createElement(Paper__default["default"], {
    elevation: 2
  }, /*#__PURE__*/React__default["default"].createElement(Box__default["default"], {
    p: 2
  }, children)))), hasToolbar && /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    item: true
  }, (hasMoveUp || hasMoveDown) && /*#__PURE__*/React__default["default"].createElement(MoveUpButton, {
    style: btnStyle,
    disabled: disabled || readonly || !hasMoveUp,
    onClick: onReorderClick(index, index - 1),
    uiSchema: uiSchema
  }), (hasMoveUp || hasMoveDown) && /*#__PURE__*/React__default["default"].createElement(MoveDownButton, {
    style: btnStyle,
    disabled: disabled || readonly || !hasMoveDown,
    onClick: onReorderClick(index, index + 1),
    uiSchema: uiSchema
  }), hasRemove && /*#__PURE__*/React__default["default"].createElement(RemoveButton, {
    style: btnStyle,
    disabled: disabled || readonly,
    onClick: onDropIndexClick(index),
    uiSchema: uiSchema
  })));
};

const ArrayFieldTemplate = props => {
  const {
    canAdd,
    disabled,
    idSchema,
    uiSchema,
    items,
    onAddClick,
    readonly,
    registry,
    required,
    schema,
    title
  } = props;
  const uiOptions = utils.getUiOptions(uiSchema);
  const ArrayFieldDescriptionTemplate = utils.getTemplate("ArrayFieldDescriptionTemplate", registry, uiOptions);
  const ArrayFieldItemTemplate = utils.getTemplate("ArrayFieldItemTemplate", registry, uiOptions);
  const ArrayFieldTitleTemplate = utils.getTemplate("ArrayFieldTitleTemplate", registry, uiOptions);
  // Button templates are not overridden in the uiSchema
  const {
    ButtonTemplates: {
      AddButton
    }
  } = registry.templates;
  return /*#__PURE__*/React__default["default"].createElement(Paper__default["default"], {
    elevation: 2
  }, /*#__PURE__*/React__default["default"].createElement(Box__default["default"], {
    p: 2
  }, /*#__PURE__*/React__default["default"].createElement(ArrayFieldTitleTemplate, {
    idSchema: idSchema,
    title: uiOptions.title || title,
    schema: schema,
    uiSchema: uiSchema,
    required: required,
    registry: registry
  }), /*#__PURE__*/React__default["default"].createElement(ArrayFieldDescriptionTemplate, {
    idSchema: idSchema,
    description: uiOptions.description || schema.description,
    schema: schema,
    uiSchema: uiSchema,
    registry: registry
  }), /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    container: true,
    key: `array-item-list-${idSchema.$id}`
  }, items && items.map(_ref => {
    let {
      key,
      ...itemProps
    } = _ref;
    return /*#__PURE__*/React__default["default"].createElement(ArrayFieldItemTemplate, {
      key: key,
      ...itemProps
    });
  }), canAdd && /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    container: true,
    justifyContent: "flex-end"
  }, /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    item: true
  }, /*#__PURE__*/React__default["default"].createElement(Box__default["default"], {
    mt: 2
  }, /*#__PURE__*/React__default["default"].createElement(AddButton, {
    className: "array-item-add",
    onClick: onAddClick,
    disabled: disabled || readonly,
    uiSchema: uiSchema
  })))))));
};

const BaseInputTemplate = _ref => {
  let {
    id,
    placeholder,
    required,
    readonly,
    disabled,
    type,
    label,
    value,
    onChange,
    onBlur,
    onFocus,
    autofocus,
    options,
    schema,
    uiSchema,
    rawErrors = [],
    formContext,
    registry,
    ...textFieldProps
  } = _ref;
  const inputProps = utils.getInputProps(schema, type, options);
  // Now we need to pull out the step, min, max into an inner `inputProps` for material-ui
  const {
    step,
    min,
    max,
    ...rest
  } = inputProps;
  const otherProps = {
    inputProps: {
      step,
      min,
      max,
      ...(schema.examples ? {
        list: `examples_${id}`
      } : undefined)
    },
    ...rest
  };
  const _onChange = _ref2 => {
    let {
      target: {
        value
      }
    } = _ref2;
    return onChange(value === "" ? options.emptyValue : value);
  };
  const _onBlur = _ref3 => {
    let {
      target: {
        value
      }
    } = _ref3;
    return onBlur(id, value);
  };
  const _onFocus = _ref4 => {
    let {
      target: {
        value
      }
    } = _ref4;
    return onFocus(id, value);
  };
  const {
    schemaUtils
  } = registry;
  const displayLabel = schemaUtils.getDisplayLabel(schema, uiSchema);
  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement(TextField__default["default"], {
    id: id,
    name: id,
    placeholder: placeholder,
    label: displayLabel ? label || schema.title : false,
    autoFocus: autofocus,
    required: required,
    disabled: disabled || readonly,
    ...otherProps,
    value: value || value === 0 ? value : "",
    error: rawErrors.length > 0,
    onChange: _onChange,
    onBlur: _onBlur,
    onFocus: _onFocus,
    ...textFieldProps
  }), schema.examples && /*#__PURE__*/React__default["default"].createElement("datalist", {
    id: `examples_${id}`
  }, schema.examples.concat(schema.default ? [schema.default] : []).map(example => {
    return /*#__PURE__*/React__default["default"].createElement("option", {
      key: example,
      value: example
    });
  })));
};

const DescriptionField = _ref => {
  let {
    id,
    description
  } = _ref;
  if (description) {
    return /*#__PURE__*/React__default["default"].createElement(Typography__default["default"], {
      id: id,
      variant: "subtitle2",
      style: {
        marginTop: "5px"
      }
    }, description);
  }
  return null;
};

const ErrorList = _ref => {
  let {
    errors
  } = _ref;
  return /*#__PURE__*/React__default["default"].createElement(Paper__default["default"], {
    elevation: 2
  }, /*#__PURE__*/React__default["default"].createElement(Box__default["default"], {
    mb: 2,
    p: 2
  }, /*#__PURE__*/React__default["default"].createElement(Typography__default["default"], {
    variant: "h6"
  }, "Errors"), /*#__PURE__*/React__default["default"].createElement(List__default["default"], {
    dense: true
  }, errors.map((error, i) => {
    return /*#__PURE__*/React__default["default"].createElement(ListItem__default["default"], {
      key: i
    }, /*#__PURE__*/React__default["default"].createElement(ListItemIcon__default["default"], null, /*#__PURE__*/React__default["default"].createElement(ErrorIcon__default["default"], {
      color: "error"
    })), /*#__PURE__*/React__default["default"].createElement(ListItemText__default["default"], {
      primary: error.stack
    }));
  }))));
};

function MuiIconButton(props) {
  const {
    icon,
    color,
    uiSchema,
    ...otherProps
  } = props;
  return /*#__PURE__*/React__default["default"].createElement(IconButton__default["default"], {
    ...otherProps,
    size: "small",
    color: color
  }, icon);
}
function MoveDownButton(props) {
  return /*#__PURE__*/React__default["default"].createElement(MuiIconButton, {
    title: "Move down",
    ...props,
    icon: /*#__PURE__*/React__default["default"].createElement(ArrowDownwardIcon__default["default"], {
      fontSize: "small"
    })
  });
}
function MoveUpButton(props) {
  return /*#__PURE__*/React__default["default"].createElement(MuiIconButton, {
    title: "Move up",
    ...props,
    icon: /*#__PURE__*/React__default["default"].createElement(ArrowUpwardIcon__default["default"], {
      fontSize: "small"
    })
  });
}
function RemoveButton(props) {
  const {
    iconType,
    ...otherProps
  } = props;
  return /*#__PURE__*/React__default["default"].createElement(MuiIconButton, {
    title: "Remove",
    ...otherProps,
    color: "secondary",
    icon: /*#__PURE__*/React__default["default"].createElement(RemoveIcon__default["default"], {
      fontSize: iconType === "default" ? "medium" : "small"
    })
  });
}

/** The `FieldErrorTemplate` component renders the errors local to the particular field
 *
 * @param props - The `FieldErrorProps` for the errors being rendered
 */
function FieldErrorTemplate(props) {
  const {
    errors = [],
    idSchema
  } = props;
  if (errors.length === 0) {
    return null;
  }
  const id = `${idSchema.$id}__error`;
  return /*#__PURE__*/React__default["default"].createElement(List__default["default"], {
    dense: true,
    disablePadding: true
  }, errors.map((error, i) => {
    return /*#__PURE__*/React__default["default"].createElement(ListItem__default["default"], {
      key: i,
      disableGutters: true
    }, /*#__PURE__*/React__default["default"].createElement(FormHelperText__default["default"], {
      id: id
    }, error));
  }));
}

/** The `FieldHelpTemplate` component renders any help desired for a field
 *
 * @param props - The `FieldHelpProps` to be rendered
 */
function FieldHelpTemplate(props) {
  const {
    idSchema,
    help
  } = props;
  if (!help) {
    return null;
  }
  const id = `${idSchema.$id}__help`;
  return /*#__PURE__*/React__default["default"].createElement(FormHelperText__default["default"], {
    id: id
  }, help);
}

const FieldTemplate = _ref => {
  let {
    id,
    children,
    classNames,
    disabled,
    displayLabel,
    hidden,
    label,
    onDropPropertyClick,
    onKeyChange,
    readonly,
    required,
    rawErrors = [],
    errors,
    help,
    rawDescription,
    schema,
    uiSchema,
    registry
  } = _ref;
  const uiOptions = utils.getUiOptions(uiSchema);
  const WrapIfAdditionalTemplate = utils.getTemplate("WrapIfAdditionalTemplate", registry, uiOptions);
  if (hidden) {
    return /*#__PURE__*/React__default["default"].createElement("div", {
      style: {
        display: "none"
      }
    }, children);
  }
  return /*#__PURE__*/React__default["default"].createElement(WrapIfAdditionalTemplate, {
    classNames: classNames,
    disabled: disabled,
    id: id,
    label: label,
    onDropPropertyClick: onDropPropertyClick,
    onKeyChange: onKeyChange,
    readonly: readonly,
    required: required,
    schema: schema,
    uiSchema: uiSchema,
    registry: registry
  }, /*#__PURE__*/React__default["default"].createElement(FormControl__default["default"], {
    fullWidth: true,
    error: rawErrors.length ? true : false,
    required: required
  }, children, displayLabel && rawDescription ? /*#__PURE__*/React__default["default"].createElement(Typography__default["default"], {
    variant: "caption",
    color: "textSecondary"
  }, rawDescription) : null, errors, help));
};

const ObjectFieldTemplate = _ref => {
  let {
    description,
    title,
    properties,
    required,
    disabled,
    readonly,
    uiSchema,
    idSchema,
    schema,
    formData,
    onAddClick,
    registry
  } = _ref;
  const uiOptions = utils.getUiOptions(uiSchema);
  const TitleFieldTemplate = utils.getTemplate("TitleFieldTemplate", registry, uiOptions);
  const DescriptionFieldTemplate = utils.getTemplate("DescriptionFieldTemplate", registry, uiOptions);
  // Button templates are not overridden in the uiSchema
  const {
    ButtonTemplates: {
      AddButton
    }
  } = registry.templates;
  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, (uiOptions.title || title) && /*#__PURE__*/React__default["default"].createElement(TitleFieldTemplate, {
    id: `${idSchema.$id}-title`,
    title: title,
    required: required,
    schema: schema,
    uiSchema: uiSchema,
    registry: registry
  }), (uiOptions.description || description) && /*#__PURE__*/React__default["default"].createElement(DescriptionFieldTemplate, {
    id: `${idSchema.$id}-description`,
    description: uiOptions.description || description,
    schema: schema,
    uiSchema: uiSchema,
    registry: registry
  }), /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    container: true,
    spacing: 2,
    style: {
      marginTop: "10px"
    }
  }, properties.map((element, index) =>
  // Remove the <Grid> if the inner element is hidden as the <Grid>
  // itself would otherwise still take up space.
  element.hidden ? element.content : /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    item: true,
    xs: 12,
    key: index,
    style: {
      marginBottom: "10px"
    }
  }, element.content)), utils.canExpand(schema, uiSchema, formData) && /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    container: true,
    justifyContent: "flex-end"
  }, /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    item: true
  }, /*#__PURE__*/React__default["default"].createElement(AddButton, {
    className: "object-property-expand",
    onClick: onAddClick(schema),
    disabled: disabled || readonly,
    uiSchema: uiSchema
  })))));
};

const SubmitButton = props => {
  const {
    submitText,
    norender,
    props: submitButtonProps
  } = utils.getSubmitButtonOptions(props.uiSchema);
  if (norender) {
    return null;
  }
  return /*#__PURE__*/React__default["default"].createElement(Box__default["default"], {
    marginTop: 3
  }, /*#__PURE__*/React__default["default"].createElement(Button__default["default"], {
    type: "submit",
    variant: "contained",
    color: "primary",
    ...submitButtonProps
  }, submitText));
};

const TitleField = _ref => {
  let {
    id,
    title
  } = _ref;
  return /*#__PURE__*/React__default["default"].createElement(Box__default["default"], {
    id: id,
    mb: 1,
    mt: 1
  }, /*#__PURE__*/React__default["default"].createElement(Typography__default["default"], {
    variant: "h5"
  }, title), /*#__PURE__*/React__default["default"].createElement(Divider__default["default"], null));
};

const WrapIfAdditionalTemplate = _ref => {
  let {
    children,
    classNames,
    disabled,
    id,
    label,
    onDropPropertyClick,
    onKeyChange,
    readonly,
    required,
    schema,
    uiSchema,
    registry
  } = _ref;
  // Button templates are not overridden in the uiSchema
  const {
    RemoveButton
  } = registry.templates.ButtonTemplates;
  const keyLabel = `${label} Key`; // i18n ?
  const additional = (utils.ADDITIONAL_PROPERTY_FLAG in schema);
  const btnStyle = {
    flex: 1,
    paddingLeft: 6,
    paddingRight: 6,
    fontWeight: "bold"
  };
  if (!additional) {
    return /*#__PURE__*/React__default["default"].createElement("div", {
      className: classNames
    }, children);
  }
  const handleBlur = _ref2 => {
    let {
      target
    } = _ref2;
    return onKeyChange(target.value);
  };
  return /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    container: true,
    key: `${id}-key`,
    alignItems: "center",
    spacing: 2,
    className: classNames
  }, /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    item: true,
    xs: true
  }, /*#__PURE__*/React__default["default"].createElement(FormControl__default["default"], {
    fullWidth: true,
    required: required
  }, /*#__PURE__*/React__default["default"].createElement(InputLabel__default["default"], null, keyLabel), /*#__PURE__*/React__default["default"].createElement(Input__default["default"], {
    defaultValue: label,
    disabled: disabled || readonly,
    id: `${id}-key`,
    name: `${id}-key`,
    onBlur: !readonly ? handleBlur : undefined,
    type: "text"
  }))), /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    item: true,
    xs: true
  }, children), /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    item: true
  }, /*#__PURE__*/React__default["default"].createElement(RemoveButton, {
    iconType: "default",
    style: btnStyle,
    disabled: disabled || readonly,
    onClick: onDropPropertyClick(label),
    uiSchema: uiSchema
  })));
};

var Templates = {
  ArrayFieldItemTemplate,
  ArrayFieldTemplate,
  BaseInputTemplate,
  ButtonTemplates: {
    AddButton,
    MoveDownButton,
    MoveUpButton,
    RemoveButton,
    SubmitButton
  },
  DescriptionFieldTemplate: DescriptionField,
  ErrorListTemplate: ErrorList,
  FieldErrorTemplate,
  FieldHelpTemplate,
  FieldTemplate,
  ObjectFieldTemplate,
  TitleFieldTemplate: TitleField,
  WrapIfAdditionalTemplate
};

const CheckboxWidget = props => {
  const {
    schema,
    id,
    value,
    disabled,
    readonly,
    label,
    autofocus,
    onChange,
    onBlur,
    onFocus
  } = props;
  // Because an unchecked checkbox will cause html5 validation to fail, only add
  // the "required" attribute if the field value must be "true", due to the
  // "const" or "enum" keywords
  const required = utils.schemaRequiresTrueValue(schema);
  const _onChange = (_, checked) => onChange(checked);
  const _onBlur = _ref => {
    let {
      target: {
        value
      }
    } = _ref;
    return onBlur(id, value);
  };
  const _onFocus = _ref2 => {
    let {
      target: {
        value
      }
    } = _ref2;
    return onFocus(id, value);
  };
  return /*#__PURE__*/React__default["default"].createElement(FormControlLabel__default["default"], {
    control: /*#__PURE__*/React__default["default"].createElement(Checkbox__default["default"], {
      id: id,
      name: id,
      checked: typeof value === "undefined" ? false : Boolean(value),
      required: required,
      disabled: disabled || readonly,
      autoFocus: autofocus,
      onChange: _onChange,
      onBlur: _onBlur,
      onFocus: _onFocus
    }),
    label: label || ""
  });
};

const selectValue = (value, selected, all) => {
  const at = all.indexOf(value);
  const updated = selected.slice(0, at).concat(value, selected.slice(at));
  // As inserting values at predefined index positions doesn't work with empty
  // arrays, we need to reorder the updated selection to match the initial order
  return updated.sort((a, b) => all.indexOf(a) > all.indexOf(b));
};
const deselectValue = (value, selected) => {
  return selected.filter(v => v !== value);
};
const CheckboxesWidget = _ref => {
  let {
    schema,
    label,
    id,
    disabled,
    options,
    value,
    autofocus,
    readonly,
    required,
    onChange,
    onBlur,
    onFocus
  } = _ref;
  const {
    enumOptions,
    enumDisabled,
    inline
  } = options;
  const _onChange = option => _ref2 => {
    let {
      target: {
        checked
      }
    } = _ref2;
    const all = enumOptions.map(_ref3 => {
      let {
        value
      } = _ref3;
      return value;
    });
    if (checked) {
      onChange(selectValue(option.value, value, all));
    } else {
      onChange(deselectValue(option.value, value));
    }
  };
  const _onBlur = _ref4 => {
    let {
      target: {
        value
      }
    } = _ref4;
    return onBlur(id, value);
  };
  const _onFocus = _ref5 => {
    let {
      target: {
        value
      }
    } = _ref5;
    return onFocus(id, value);
  };
  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement(FormLabel__default["default"], {
    required: required,
    htmlFor: id
  }, label || schema.title), /*#__PURE__*/React__default["default"].createElement(FormGroup__default["default"], {
    id: id,
    row: !!inline
  }, Array.isArray(enumOptions) && enumOptions.map((option, index) => {
    const checked = value.indexOf(option.value) !== -1;
    const itemDisabled = Array.isArray(enumDisabled) && enumDisabled.indexOf(option.value) !== -1;
    const checkbox = /*#__PURE__*/React__default["default"].createElement(Checkbox__default["default"], {
      id: `${id}-${option.value}`,
      name: id,
      checked: checked,
      disabled: disabled || itemDisabled || readonly,
      autoFocus: autofocus && index === 0,
      onChange: _onChange(option),
      onBlur: _onBlur,
      onFocus: _onFocus
    });
    return /*#__PURE__*/React__default["default"].createElement(FormControlLabel__default["default"], {
      control: checkbox,
      key: option.value,
      label: option.label
    });
  })));
};

const DateWidget = props => {
  const {
    options,
    registry
  } = props;
  const BaseInputTemplate = utils.getTemplate("BaseInputTemplate", registry, options);
  return /*#__PURE__*/React__default["default"].createElement(BaseInputTemplate, {
    type: "date",
    InputLabelProps: {
      shrink: true
    },
    ...props
  });
};

const DateTimeWidget = props => {
  const {
    options,
    registry
  } = props;
  const BaseInputTemplate = utils.getTemplate("BaseInputTemplate", registry, options);
  const value = utils.utcToLocal(props.value);
  const onChange = value => {
    props.onChange(utils.localToUTC(value));
  };
  return /*#__PURE__*/React__default["default"].createElement(BaseInputTemplate, {
    type: "datetime-local",
    InputLabelProps: {
      shrink: true
    },
    ...props,
    value: value,
    onChange: onChange
  });
};

const RadioWidget = _ref => {
  let {
    id,
    schema,
    options,
    value,
    required,
    disabled,
    readonly,
    label,
    onChange,
    onBlur,
    onFocus
  } = _ref;
  const {
    enumOptions,
    enumDisabled
  } = options;
  const _onChange = (_, value) => onChange(schema.type == "boolean" ? value !== "false" : value);
  const _onBlur = _ref2 => {
    let {
      target: {
        value
      }
    } = _ref2;
    return onBlur(id, value);
  };
  const _onFocus = _ref3 => {
    let {
      target: {
        value
      }
    } = _ref3;
    return onFocus(id, value);
  };
  const row = options ? options.inline : false;
  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement(FormLabel__default["default"], {
    required: required,
    htmlFor: id
  }, label || schema.title), /*#__PURE__*/React__default["default"].createElement(RadioGroup__default["default"], {
    id: id,
    name: id,
    value: `${value}`,
    row: row,
    onChange: _onChange,
    onBlur: _onBlur,
    onFocus: _onFocus
  }, Array.isArray(enumOptions) && enumOptions.map(option => {
    const itemDisabled = Array.isArray(enumDisabled) && enumDisabled.indexOf(option.value) !== -1;
    const radio = /*#__PURE__*/React__default["default"].createElement(FormControlLabel__default["default"], {
      control: /*#__PURE__*/React__default["default"].createElement(Radio__default["default"], {
        name: id,
        id: `${id}-${option.value}`,
        color: "primary"
      }),
      label: `${option.label}`,
      value: `${option.value}`,
      key: option.value,
      disabled: disabled || itemDisabled || readonly
    });
    return radio;
  })));
};

const RangeWidget = _ref => {
  let {
    value,
    readonly,
    disabled,
    onBlur,
    onFocus,
    options,
    schema,
    onChange,
    required,
    label,
    id
  } = _ref;
  const sliderProps = {
    value,
    label,
    id,
    name: id,
    ...utils.rangeSpec(schema)
  };
  const _onChange = (_, value) => {
    onChange(value ? options.emptyValue : value);
  };
  const _onBlur = _ref2 => {
    let {
      target: {
        value
      }
    } = _ref2;
    return onBlur(id, value);
  };
  const _onFocus = _ref3 => {
    let {
      target: {
        value
      }
    } = _ref3;
    return onFocus(id, value);
  };
  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement(FormLabel__default["default"], {
    required: required,
    htmlFor: id
  }, label || schema.title), /*#__PURE__*/React__default["default"].createElement(Slider__default["default"], {
    disabled: disabled || readonly,
    onChange: _onChange,
    onBlur: _onBlur,
    onFocus: _onFocus,
    valueLabelDisplay: "auto",
    ...sliderProps
  }));
};

const SelectWidget = _ref => {
  let {
    schema,
    id,
    options,
    label,
    required,
    disabled,
    readonly,
    value,
    multiple,
    autofocus,
    onChange,
    onBlur,
    onFocus,
    rawErrors = []
  } = _ref;
  const {
    enumOptions,
    enumDisabled
  } = options;
  const emptyValue = multiple ? [] : "";
  const _onChange = _ref2 => {
    let {
      target: {
        value
      }
    } = _ref2;
    return onChange(utils.processSelectValue(schema, value, options));
  };
  const _onBlur = _ref3 => {
    let {
      target: {
        value
      }
    } = _ref3;
    return onBlur(id, utils.processSelectValue(schema, value, options));
  };
  const _onFocus = _ref4 => {
    let {
      target: {
        value
      }
    } = _ref4;
    return onFocus(id, utils.processSelectValue(schema, value, options));
  };
  return /*#__PURE__*/React__default["default"].createElement(TextField__default["default"], {
    id: id,
    name: id,
    label: label || schema.title,
    select: true,
    value: typeof value === "undefined" ? emptyValue : value,
    required: required,
    disabled: disabled || readonly,
    autoFocus: autofocus,
    error: rawErrors.length > 0,
    onChange: _onChange,
    onBlur: _onBlur,
    onFocus: _onFocus,
    InputLabelProps: {
      shrink: true
    },
    SelectProps: {
      multiple: typeof multiple === "undefined" ? false : multiple
    }
  }, enumOptions.map((_ref5, i) => {
    let {
      value,
      label
    } = _ref5;
    const disabled = enumDisabled && enumDisabled.indexOf(value) != -1;
    return /*#__PURE__*/React__default["default"].createElement(MenuItem__default["default"], {
      key: i,
      value: value,
      disabled: disabled
    }, label);
  }));
};

const TextareaWidget = props => {
  const {
    options,
    registry
  } = props;
  const BaseInputTemplate = utils.getTemplate("BaseInputTemplate", registry, options);
  let rows = 5;
  if (typeof options.rows === "string" || typeof options.rows === "number") {
    rows = options.rows;
  }
  return /*#__PURE__*/React__default["default"].createElement(BaseInputTemplate, {
    ...props,
    multiline: true,
    rows: rows
  });
};

var Widgets = {
  CheckboxWidget,
  CheckboxesWidget,
  DateWidget,
  DateTimeWidget,
  RadioWidget,
  RangeWidget,
  SelectWidget,
  TextareaWidget
};

const Theme = {
  templates: Templates,
  widgets: Widgets
};

const MuiForm = /*#__PURE__*/core.withTheme(Theme);

exports.Form = MuiForm;
exports.Templates = Templates;
exports.Theme = Theme;
exports.Widgets = Widgets;
exports["default"] = MuiForm;
